src/pip/_internal/req/constructors.py

```diff
@@ -504,3 +505,45 @@ def install_req_from_link_and_ireq(
         config_settings=ireq.config_settings,
         user_supplied=ireq.user_supplied,
     )
+
+
+def install_req_drop_extras(ireq: InstallRequirement) -> InstallRequirement:
+    """
+    Creates a new InstallationRequirement using the given template but without
+    any extras. Sets the original requirement as the new one's parent
+    (comes_from).
+    """
+    req = Requirement(str(ireq.req))
+    req.extras = {}
```

Comment:
Instead of trying to modify a Requirement in-place (which IIRC is not documented to be an officially supported operation), it may be better to strip the extras in the requirement string instead. This should not be too dificult with some regex since both the project and extra names are quite predictable.

Reply:
I made the change though I don't like the fact that it feels a bit more brittle. On the other hand I do agree with your sentiment that I shouldn't something that isn't part of `packaging.Requirement`'s contract. @pradyunsg do you have an opinion on this (faa3289a9)?

Reply:
I'd be happy if `packaging` made it a supported operation, but the existing docs state that the `Requirement` class is for *parsing* the string form of a requirement, which to me doesn't imply the result is mutable. You *could* read it as saying that it parses the string into a structured form, which you are allowed to them mutate if you want, but I don't think that automatically follows.

So as things stand, I agree with @uranusjr.

As @pradyunsg is a `packaging` maintainer, if he says it's OK, then I'm fine with that (although I'd ask that the docs be changed to explicitly state what's supported).

Reply:
Yea, don't modify this in place. It doesn't break things as implemented right now but making this a frozen dataclass is something I want to do in the future.

Reply:
Thanks for your input. A frozen dataclass would be nice, that would then allow for `dataclasses.replace`, which I'd always prefer over in-place modifications.